src/domain/service/fileUploader.service.ts

```diff
@@ -0,0 +1,238 @@
+
+/**
+ * File data for upload
+ */
+interface FileToUpload {
+  /**
+   * File data
+   */
+  data: FileData;
+  /**
+   * File name
+   */
+  name: string;
+  /**
+   * Mimetype of the file
+   */
+  mimetype: string;
+  /**
+   * File type
+   */
+  type: FileTypes;
+}
+
+/**
+ * File upload details, contains user id, who uploaded the file and note id, in case if file is a part of note
+ */
+interface FileUploadDetails {
+  /**
+   * User who uploaded the file
+   */
+  userId?: User['id'];
+
+  /**
+   * In case if file is a part of note, note id to identify permissions to access
+   */
+  noteId?: NoteInternalId;
+}
+
+/**
+ * File uploader service
+ */
+export default class FileUploaderService {
+  /**
+   * Repository for objects
+   */
+  private readonly objectRepository: ObjectRepository;
+
+  /**
+   * Repository for files data
+   */
+  private readonly fileRepository: FileRepository;
+
+  /**
+   * Note settings shared methods
+   */
+  private readonly noteSettingsSharedMethods: NoteSettingsSharedMethods;
+
+  /**
+   * File uploader service constructor
+   *
+   * @param objectRepository - repository for objects
+   * @param fileRepository - repository for files data
+   * @param noteSettingsSharedMethods - note settings shared methods
+   */
+  constructor(objectRepository: ObjectRepository, fileRepository: FileRepository, noteSettingsSharedMethods: NoteSettingsSharedMethods) {
+    this.noteSettingsSharedMethods = noteSettingsSharedMethods;
+    this.objectRepository = objectRepository;
+    this.fileRepository = fileRepository;
+  }
+
+  /**
+   * Upload file
+   *
+   * @param fileData - file data to upload (e.g. buffer, name, mimetype)
+   * @param details - file upload details (e.g. user id, note id)
+   */
+  public async uploadFile(fileData: FileToUpload, details?: FileUploadDetails): Promise<string> {
+    const key = createFileId();
+
+    let hasPermissions: boolean;
+
+    /**
+     * If file is a part of note, check if user has permissions to access file
+     */
+    if (details?.noteId !== undefined) {
+      hasPermissions = await this.hasPermissionsToAccessFileInNote(details.noteId, MemberRole.Write, details.userId);
+    } else {
+      hasPermissions = true;
+    }
+
+    if (hasPermissions === false) {
+      throw new DomainError('User has no permissions to access file');
+    }
+
+    const bucket = this.defineBucketByFileType(fileData.type);
+
+    const uploaded = await this.objectRepository.insert(fileData.data, key, bucket);
+
+    if (uploaded === null) {
+      throw new DomainError('File was not uploaded');
+    }
+
+    const file = await this.fileRepository.insert({
+      ...fileData,
+      key,
+      userId: details?.userId,
+      noteId: details?.noteId,
+      size: fileData.data.length,
+    });
+
+    if (file === null) {
+      throw new Error('File was not uploaded');
+    }
+
+    return file.key;
+  }
+
+  /**
+   * Get file data by key
+   *
+   * @param key - file key
```

Comment:
name and description is unclear. What key? It's better to use 2-words names for such generic words like `key, data` etc

Reply:
renamed, updated docs